File: PyGameTutorial/GameField.py
import pygame


class FieldBlock(object):
    def __init__(self,x,y,size):
        self.R = pygame. Rect(x,y,size,size)
        self.sprite=None

class GameField(object):
    def __init__(self,sx,sy):
        self.F=[];
        self.sx=sx
        self.sy=sy
        self.blocksize=16
        for y in range(24):
            self.F.append([]);
            for x in range(10):
                self.F[y].append(FieldBlock((x* self.blocksize)+sx,
                                            (y* self.blocksize)+sy, self.blocksize))
    def printField(self):
        i =0
        for r in self.F:
            line=str(i)+":"
            for x in r:
                if(x.sprite is not None):
                    line += "X,"
                    #str(x.sprite.frames[0].x)+
                else:
                    line+= " ,"
            print(line)
            i+=1

    def ClearLine(self,rowI):
            row = self.F[rowI]
            for x in range(0,len(row)):
                row[x].sprite=None

    def ClearLines(self,lineDown):
        self.printField()
        for rowI in lineDown:
            self.ClearLine(rowI)
        self.printField()
 

    def DragLineDownTo(self,rowI):
           
        for r in reversed(range(1,rowI+1)):
            y =  self.sy + (r*self.blocksize)
            for x in range(0,len(self.F[r])):
                self.F[r][x].sprite = self.F[r-1][x].sprite
                if self.F[r][x].sprite is not None:
                    self.F[r][x].sprite.R.y = y
        
        self.ClearLine(0)

    def GetLineDown(self,block):
        
        linesDown =[]
        yI = []
        for s in block.sprites:
            relY = s.R.y-self.sy
            try:
                yI.index(relY) 
            except ValueError:
                yI.append(relY)
                print("checkline ", yI)

        for rowI in yI:#reversed( yI):
            count=0
            row = self.F[int(rowI/16)]
            for fb in row:
                if fb.sprite is not None:
                    count+=1
            if count == len(row):
                print("linecomplete")
                linesDown.append( self.F.index(row))

        print(linesDown)
        return linesDown
    def Overlaps(self, block):

        for s in block.sprites:
            for row in self.F:
                for fb in row:
                    if fb.sprite is not None and fb.R.x == s.R.x and fb.R.y == s.R.y:
                        return True
        return False
    
    def MovedOutsideVerti(self,block):
        for s in block.sprites:
            if s.R.bottom > self.F[len(self.F)-1][0].R.bottom:
                return True
        return False
    def MovedOutsideHori(self,block):
        for s in block.sprites:
            if s.R.left <self.F[0][0].R.x:
               return True
            if s.R.x >= self.F[0][len(self.F[0])-1].R.right:
               return True
        return False
    
    def Rotate(left):

        pass
    def PutBlock(self, s):
        
        for row in self.F:
            for fb in row:
                if fb.R.x == s.R.x and fb.R.y == s.R.y:
                    fb.sprite = s
                    return

File: js/GameField.js


class FieldBlock
{
     constructor(x,y,size)
     {
        this.R = new Rect(x,y,size,size)
        this.sprite=null
     }
}

class GameField
{
     constructor(sx,sy)
     {
        this.F=[];
        this.sx=sx
        this.sy=sy
        this.blocksize=16
        for (let y = 0 ; y < 24;y++)
        {
            this.F.push([]);
            for (let x =0 ; x <  10;x++)
            {
                let b=new FieldBlock((x* this.blocksize)+sx, (y* this.blocksize)+sy, this.blocksize)
                this.F[y].push(b)
            }
        }
    }

    printField()
    {
        let   i =0
        for (let r in this.F)
        {
            r= this.F[r];

            let line=i+"{"
            for (let x in r)
            {
                x = r[x];
                if(x.sprite !=null)
                {
                    line += "X,"
                    //#str(x.sprite.frames[0].x)+
                }
                else
                {
                    line+= " ,"
                }
            }
            console.log(line)
            i+=1
        }
    }

     ClearLine(rowI)
     {
            let    row = this.F[rowI]
            for (let x = 0 ; x < row.length;x++)
            {
                row[x].sprite=null
            }
    }
     ClearLines(lineDown)
     {
        this.printField()
        for (let rowI in lineDown)
        {
            rowI = lineDown[rowI]
            this.ClearLine(lineDown[rowI])
        }
        this.printField()
    }

     DragLineDownTo(rowI)
     {
           
        //for( let r = rowI+1;)
        for (let r = rowI; r > 1; r--)
        //for (let r in reversed(range(1,rowI+1)))
        {
            let y = this.sy + (r*this.blocksize)
            for(let  x = 0; x< this.F[r].length; x++)
            {
                this.F[r][x].sprite = this.F[r-1][x].sprite
                if( this.F[r][x].sprite != null)
                {
                    this.F[r][x].sprite.R.y = y
                }
            }
        }
        this.ClearLine(0)
    }
     GetLineDown(block)
     {
        
        let linesDown =[]
        let  yI = []
        for (let s in block.sprites)
        {
            s=  block.sprites[s];

            let   relY = s.R.y-this.sy
            try{
                yI.index(relY)
            }
            catch 
            //except ValueError
            {
                yI.push(relY)
              //  print("checkline ", yI)
            }
        }
        for (let rowI in yI)//{#reversed( yI){
        {
            rowI= yI[rowI];
            let  count=0
            let   row = this.F[Math.floor(rowI/16)]
            for( let fb in row)
            {
                fb= row[fb];

                if (fb.sprite != null)
                {
                    count+=1
                }
            }
            if( count == row.length)
            {
                console.log("linecomplete")
                let index=this.F.indexOf(row);
                linesDown.push(index);
                //linesDown.append( this.F.index(row))
            }
        }
        console.log(linesDown)
        return linesDown
    }
    Overlaps( block)
    {

        for (let s in block.sprites)
        {
            s = block.sprites[s];

            for (let row in this.F)
            {
                row = this.F[row]
                for( let fb in row)
                {
                    fb = row[fb]
                    if (fb.sprite != null && fb.R.x == s.R.x && fb.R.y == s.R.y)
                    {
                        return true
                    }
                }
            }
        }
        return false
    }
   
     MovedOutsideVerti(block){
        for( let s in block.sprites)
        {
            s = block.sprites[s]
            if (s.R.bottom() > this.F[this.F.length-1][0].R.bottom())
            {
                return true
            }
        }
        return false
     }
     MovedOutsideHori(block)
     {
        for (let s in block.sprites)
        { 
            s =block.sprites[s]
            if( s.R.left() <this.F[0][0].R.x)
            {
               return true
            }
            if( s.R.x >= this.F[0][this.F[0].length-1].R.right())
            {
               return true
            }
        }
        return false
     }
     Rotate(left)
     {

      //  pass
     }
     PutBlock( s)
     {
        
        for( let  row in this.F)
        {
            row = this.F[row]
            for (let fb in row)
            {
                fb = row[fb]
                if (fb.R.x == s.R.x && fb.R.y == s.R.y)
                {
                    fb.sprite = s
                    return
                }
            }
        }
    }
}